File: binary_search/boj/p2805.py
#pypy로 제출하니까 성공
from sys import stdin

n, m = map(int, stdin.readline().split())
tree = list(map(int, stdin.readline().split()))

start = 0
end = max(tree)
result = 0

while start <= end:
    sum = 0
    mid = (start + end) // 2

    for i in tree:
        if i > mid:
            sum += i - mid
    
    if sum < m:
        end = mid - 1
    else:
        result = mid
        start = mid + 1

print(result)

File: binary_search/boj/p2805.java
package binary_search.boj;
import java.util.Scanner;
/**
 * p2805
 */
public class p2805 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        int[] tree = new int[n];
    
        for (int i = 0; i < tree.length; i++) {
          tree[i] = sc.nextInt();
        }
    
        int start = 0;
        int mid = 0;
        long res = 0;
        long sum = 0;
        int end = (int)1e9;
    
        while (start <= end) {
          sum = 0;
          mid = (start + end) / 2;
    
          for (int i: tree) {
            if (i > mid) sum += (i - mid);
          }
    
          if (sum < m) {
            end = mid - 1;
          } else {
            res = mid;
            start = mid + 1;
          }
        }
    
        System.out.println(res);
        sc.close();
      }

    
}